torch/optim/swa_utils.py

```diff
@@ -0,0 +1,267 @@
+            equally weighted average is used (default: None)
+
+    Example:
+        >>> loader, optimizer, model, loss_fn = ...
+        >>> swa_model = torch.optim.swa_utils.AveragedModel(model)
+        >>> scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, 
+        >>>                                     T_max=300)
+        >>> swa_start = 160
+        >>> swa_scheduler = SWALR(optimizer, swa_lr=0.05)
+        >>> for i in range(300):
+        >>>      for input, target in loader:
+        >>>          optimizer.zero_grad()
+        >>>          loss_fn(model(input), target).backward()
+        >>>          optimizer.step()
+        >>>      if i > swa_start:
+        >>>          swa_model.update_parameters(model)
+        >>>          swa_scheduler.step()
+        >>>      else:
+        >>>          scheduler.step()
+        >>>
+        >>> # Update bn statistics for the swa_model at the end
+        >>> torch.optim.swa_utils.update_bn(loader, swa_model) 
+
+    You can use custom averaging functions with `avg_fn` parameter.
+    If no averaging function is provided, the default is to compute
+    equally-weighted average of the weights.
+
+    Example:
+        >>> # ema_model computes exponential moving averages of the weights
+        >>> ema_avg = lambda averaged_model_parameter, model_parameter, num_averaged:\
+                            0.1 * averaged_model_parameter + 0.9 * model_parameter
+        >>> ema_model = torch.optim.swa_utils.AveragedModel(model, avg_fn=ema_avg)
+
+    .. note::
+        When using SWA with models containing Batch Normalization you may 
+        need to update the activation statistics for Batch Normalization.
+        You can do so by using :meth:`torch.optim.swa_utils.update_bn` utility.
+
+    .. note::
+        :attr:`avg_fn` is not saved in the :meth:`state_dict` of the model.
+
+    .. note::
+        When :meth:`update_parameters` is called for the first time (i.e. 
+        :attr:`n_averaged` is `0`) the parameters of `model` are copied
+        to the parameters of :class:`AveragedModel`. For every subsequent
+        call of :meth:`update_parameters` the function `avg_fn` is used
+        to update the parameters.
+
+    .. _Averaging Weights Leads to Wider Optima and Better Generalization:
+        https://arxiv.org/abs/1803.05407
+    .. _There Are Many Consistent Explanations of Unlabeled Data: Why You Should
+        Average:
+        https://arxiv.org/abs/1806.05594
+    .. _SWALP: Stochastic Weight Averaging in Low-Precision Training:
+        https://arxiv.org/abs/1904.11943
+    .. _Stochastic Weight Averaging in Parallel: Large-Batch Training That 
+        Generalizes Well:
+        https://arxiv.org/abs/2001.02312
+    """
+    def __init__(self, model, device=None, avg_fn=None):
+        super(AveragedModel, self).__init__()
+        self.module = deepcopy(model)
+        if device is not None:
+            self.module = self.module.to(device)
```

Comment:
The model could have layers on different devices (e.g. model parallel, where a model's layers are split across more than one gpu). 

I'm assuming you may want to move the model to avoid having the gpu run out of memory, is that correct?

Reply:
Right, that's the idea. We don't have to use the `AvergedModel` parameters during training, and we can even store them on a CPU. 

If `device` is not specified, the `AvergedModel` parameters are stored on the same devices as the parameters of the original model, and it works if the layers are on different devices. I have a small test for this case here: https://github.com/pytorch/pytorch/blob/41ad473eba04ab5118003f29f161d08943cf83af/test/test_optim.py#L1703-L1724